README.md
# pyproject-validate

| | |
| --- | --- |
| CI/CD | [![CI - Test](https://github.com/ofek/pyproject-validate/actions/workflows/test.yml/badge.svg)](https://github.com/ofek/pyproject-validate/actions/workflows/test.yml) [![CD - Build](https://github.com/ofek/pyproject-validate/actions/workflows/build.yml/badge.svg)](https://github.com/ofek/pyproject-validate/actions/workflows/build.yml) |
| Package | [![PyPI - Version](https://img.shields.io/pypi/v/pyproject-validate.svg?logo=pypi&label=PyPI&logoColor=gold)](https://pypi.org/project/pyproject-validate/) [![PyPI - Downloads](https://img.shields.io/pypi/dm/pyproject-validate.svg?color=blue&label=Downloads&logo=pypi&logoColor=gold)](https://pypi.org/project/pyproject-validate/) [![PyPI - Python Version](https://img.shields.io/pypi/pyversions/pyproject-validate.svg?logo=python&label=Python&logoColor=gold)](https://pypi.org/project/pyproject-validate/) |
| Meta | [![code style - black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black) [![types - Mypy](https://img.shields.io/badge/types-Mypy-blue.svg)](https://github.com/ambv/black) [![imports - isort](https://img.shields.io/badge/imports-isort-ef8336.svg)](https://github.com/pycqa/isort) [![License - MIT](https://img.shields.io/badge/license-MIT-9400d3.svg)](https://spdx.org/licenses/) [![GitHub Sponsors](https://img.shields.io/github/sponsors/ofek?logo=GitHub%20Sponsors&style=social)](https://github.com/sponsors/ofek) |

-----

Validate and format `pyproject.toml` files.

**Table of Contents**

- [Installation](#installation)
- [Usage](#usage)
- [Validators](#validators)
  - [Specs](#specs)
  - [Naming](#naming)
  - [Dependencies](#dependencies)
- [License](#license)

## Installation

```console
pip install pyproject-validate
```

## Usage

```console
usage: pyproject-validate [-h] [--fix] [--config CONFIG] [--version]

optional arguments:
  -h, --help       show this help message and exit
  --fix            whether to apply fixes for any encountered errors
  --config CONFIG  explicit path to the project config file
  --version        show program's version number and exit
```

## Validators

### Specs

Adhere to the data model defined by [PEP 517](https://www.python.org/dev/peps/pep-0517/#source-trees) and [PEP 621](https://www.python.org/dev/peps/pep-0621/#details).

### Naming

Ensure normalized project names.

Before:

```toml
[project]
name = "Foo.bAr"
```

After:

```toml
name = "foo-bar"
```

### Dependencies

Ensure normalized and sorted [PEP 508](https://www.python.org/dev/peps/pep-0508/) dependency definitions.

Before:

```toml
[project]
dependencies = [
    "python-dateutil",
    "bAr.Baz[TLS]   >=1.2RC5",
    'Foo;python_version<"3.8"',
]
```

After:

```toml
dependencies = [
    "bar-baz[tls]>=1.2rc5",
    "foo; python_version < '3.8'",
    "python-dateutil",
]
```

## License

`pyproject-validate` is distributed under the terms of the [MIT](https://spdx.org/licenses/MIT.html) license.


HISTORY.md
# History

-----

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 0.1.0 - 2022-02-21

This is the initial public release.


tests/conftest.py
import os
import sys
from collections import namedtuple
from pathlib import Path
from tempfile import TemporaryDirectory

import pytest

from pyproject_validate.cli import main

InvocationResult = namedtuple("InvocationResult", ["code", "output"])


class ProjectFile:
    def __init__(self, directory: Path):
        self.directory = directory
        self.path = directory / "pyproject.toml"

    def read(self) -> str:
        return self.path.read_text(encoding="utf-8")

    def write(self, text: str):
        self.path.write_text(text, encoding="utf-8")


def _invoke(args, capsys):
    with pytest.raises(SystemExit) as e:
        original_sys_argv = sys.argv
        try:
            sys.argv = ["pyproject-validate", *args]
            main()
        finally:
            sys.argv = original_sys_argv

    return InvocationResult(e.value.code, "".join(capsys.readouterr()))


@pytest.fixture
def invoke(capsys):
    return lambda *args: _invoke(args, capsys)


@pytest.fixture
def project_file():
    with TemporaryDirectory() as d:
        path = Path(d).resolve()
        origin = os.getcwd()

        try:
            os.chdir(path)
            yield ProjectFile(path)
        finally:
            os.chdir(origin)


tests/test_name.py
class TestInvalidCharacters:
    BEFORE = """\
[build-system]
requires = [
    "hatchling",
]
build-backend = "hatchling.build"

[project]
name = "foo bar"
version = "0.0.1"
"""

    def test_error(self, project_file, invoke):
        project_file.write(self.BEFORE)

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< naming >>>
error: must only contain ASCII letters/digits, underscores, hyphens, and periods
"""
        )

    def test_cannot_fix(self, project_file, invoke):
        project_file.write(self.BEFORE)

        result = invoke("--fix")

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< naming >>>
error: must only contain ASCII letters/digits, underscores, hyphens, and periods
"""
        )


class TestNormalization:
    BEFORE = """\
[build-system]
requires = [
    "hatchling",
]
build-backend = "hatchling.build"

[project]
name = "Foo.bAr"
version = "0.0.1"
"""
    AFTER = """\
[build-system]
requires = [
    "hatchling",
]
build-backend = "hatchling.build"

[project]
name = "foo-bar"
version = "0.0.1"
"""

    def test_error(self, project_file, invoke):
        project_file.write(self.BEFORE)

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< naming >>>
error: should be foo-bar
"""
        )

    def test_fix(self, project_file, invoke):
        project_file.write(self.BEFORE)

        result = invoke("--fix")

        assert result.code == 0, result.output
        assert not result.output
        assert project_file.read() == self.AFTER

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output


tests/test_spec.py
class TestBuildSystem:
    def test_requires_missing(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for BuildSystemConfig
requires
  field required (type=value_error.missing)
"""
        )

    def test_requires_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = "hatchling"
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for BuildSystemConfig
requires
  value is not a valid list (type=type_error.list)
"""
        )

    def test_build_backend_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = []

[project]
name = "foo"
version = "0.0.1"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for BuildSystemConfig
build-backend
  str type expected (type=type_error.str)
"""
        )

    def test_backend_path_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
backend-path = "."

[project]
name = "foo"
version = "0.0.1"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for BuildSystemConfig
backend-path
  value is not a valid list (type=type_error.list)
"""
        )


class TestAuthors:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
authors = [
    { name = "U.N. Owen", email = "void@some.where" },
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
authors = [
    { name = [9000], email = [] },
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 2 validation errors for ProjectConfig
authors -> 0 -> email
  str type expected (type=type_error.str)
authors -> 0 -> name
  str type expected (type=type_error.str)
"""
        )


class TestClassifiers:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
classifiers = [
    "Development Status :: 4 - Beta",
    "Programming Language :: Python :: Implementation :: CPython",
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
classifiers = [
    "Development Status :: 4 - Beta",
    [],
    "Programming Language :: Python :: Implementation :: CPython",
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
classifiers -> 1
  str type expected (type=type_error.str)
"""
        )


class TestDependencies:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
dependencies = [
    "bar",
    "foo",
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
dependencies = [
    "foo",
    [],
    "bar",
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
dependencies -> 1
  str type expected (type=type_error.str)
"""
        )


class TestDescription:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
description = "foo"
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
description = []
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
description
  str type expected (type=type_error.str)
"""
        )


class TestDynamic:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
dynamic = [
    "foo",
    "bar",
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
dynamic = [
    "foo",
    [],
    "bar",
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
dynamic -> 1
  str type expected (type=type_error.str)
"""
        )

    def test_name(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
dynamic = [
    "name",
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
dynamic
  the `name` field must not be listed as dynamic (type=value_error)
"""
        )

    def test_redefined(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
description = "foo"
dynamic = [
    "description",
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
__root__
  field(s) defined but also listed as dynamic: description (type=value_error)
"""
        )


class TestEntryPoints:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.entry-points.foo]
bar = "baz"
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.entry-points.foo]
bar = []
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
entry-points -> foo -> bar
  str type expected (type=type_error.str)
"""
        )


class TestGUIScripts:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.gui-scripts]
foo = "bar"
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.gui-scripts]
foo = []
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
gui-scripts -> foo
  str type expected (type=type_error.str)
"""
        )


class TestKeywords:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
keywords = [
    "foo",
    "bar",
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
keywords = [
    "foo",
    [],
    "bar",
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
keywords -> 1
  str type expected (type=type_error.str)
"""
        )


class TestLicense:
    def test_valid_string(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
license = "foo"
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_valid_table(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
license = { file = "LICENSE.txt" }
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
license = { file = "LICENSE.txt", text = "accepted" }
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 2 validation errors for ProjectConfig
license
  str type expected (type=type_error.str)
license -> __root__
  cannot contain both a `file` and `text` field (type=value_error)
"""
        )


class TestLicenseFiles:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
license-files = { paths = ["LICENSE.txt"] }
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
license-files = { paths = "LICENSE.txt", globs = "LICENSES/*" }
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 2 validation errors for ProjectConfig
license-files -> globs
  value is not a valid list (type=type_error.list)
license-files -> paths
  value is not a valid list (type=type_error.list)
"""
        )

    def test_empty(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
license-files = {}
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
license-files -> __root__
  must contain either a `globs` or `paths` field (type=value_error)
"""
        )


class TestMaintainers:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
maintainers = [
    { name = "U.N. Owen", email = "void@some.where" },
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
maintainers = [
    { name = [9000], email = [] },
]
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 2 validation errors for ProjectConfig
maintainers -> 0 -> email
  str type expected (type=type_error.str)
maintainers -> 0 -> name
  str type expected (type=type_error.str)
"""
        )


class TestName:
    def test_missing(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
version = "0.0.1"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
name
  field required (type=value_error.missing)
"""
        )

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = []
version = "0.0.1"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
name
  str type expected (type=type_error.str)
"""
        )


class TestOptionalDependencies:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.optional-dependencies]
foo = [
    "bar",
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.optional-dependencies]
foo = "bar"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
optional-dependencies -> foo
  value is not a valid list (type=type_error.list)
"""
        )


class TestReadme:
    def test_valid_string(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
readme = "README.md"
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_valid_table(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
readme = { file = "README", content-type = "text/plain" }
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
readme = { file = "README.md", text = "docs" }
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 2 validation errors for ProjectConfig
readme
  str type expected (type=type_error.str)
readme -> __root__
  cannot contain both a `file` and `text` field (type=value_error)
"""
        )

    def test_invalid_content_type(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
readme = { file = "README.md", content-type = "foo" }
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 2 validation errors for ProjectConfig
readme
  str type expected (type=type_error.str)
readme -> content-type
  must be one of: text/markdown, text/x-rst, text/plain (type=value_error)
"""
        )

    def test_invalid_extension(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"
readme = "README"
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
readme
  must have one of the following extensions: .md, .rst, .txt (type=value_error)
"""
        )


class TestScripts:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.scripts]
foo = "bar"
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.scripts]
foo = []
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
scripts -> foo
  str type expected (type=type_error.str)
"""
        )


class TestURLs:
    def test_valid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.urls]
foo = "bar"
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[project.urls]
foo = []
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 1 validation error for ProjectConfig
urls -> foo
  str type expected (type=type_error.str)
"""
        )


class TestVersion:
    def test_valid_dynamic(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
dynamic = [
    "version",
]
"""
        )

        result = invoke()

        assert result.code == 0, result.output
        assert not result.output

    def test_invalid(self, project_file, invoke):
        project_file.write(
            """\
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = []
"""
        )

        result = invoke()

        assert result.code == 1, result.output
        assert (
            result.output
            == """\
<<< specs >>>
error: 2 validation errors for ProjectConfig
version
  str type expected (type=type_error.str)
__root__
  missing field(s): version (type=value_error)
"""
        )


src/pyproject_validate/cli.py
import argparse
import sys

from ._version import version
from .handlers import get_handler
from .validators import get_validators


def main():
    parser = argparse.ArgumentParser(prog="pyproject-validate", allow_abbrev=False)
    parser.add_argument("--fix", action="store_true", help="whether to apply fixes for any encountered errors")
    parser.add_argument("--config", help="explicit path to the project config file")
    if sys.version_info[:2] >= (3, 8):
        parser.add_argument("--version", action="version", version=f"%(prog)s {version}")
    args = parser.parse_args()

    handler = get_handler(args.config)
    try:
        data = handler.load()
    except Exception as e:
        print(e)
        sys.exit(1)

    errors_occurred = False
    unfixable_errors = False
    need_fixing = False
    for name, validator in get_validators().items():
        errors = []
        warnings = []
        validator.validate(data, errors, warnings)

        if (errors or warnings) and not args.fix or not validator.fixable:
            print(f"<<< {name} >>>")
            for error in errors:
                print(f"error: {error}")
            for warning in warnings:
                print(f"warning: {warning}")

        if errors:
            errors_occurred = True
            if args.fix and validator.fixable:
                need_fixing = True
                validator.fix(data)
            else:
                unfixable_errors = True
                if validator.exit_early:
                    break

    if need_fixing and not unfixable_errors:
        handler.save(data)
        errors_occurred = False

    sys.exit(int(errors_occurred))


src/pyproject_validate/handlers.py
from __future__ import annotations

import os
from abc import ABC, abstractmethod
from typing import Any, Dict, Optional


class Handler(ABC):
    def __init__(self, path: Optional[str] = None):
        self._path = path

    @property
    def path(self):
        if self._path is None:
            root = os.getcwd()
            while True:
                path = os.path.join(root, "pyproject.toml")
                if os.path.isfile(path):
                    self._path = path
                    break

                new_root = os.path.dirname(root)
                if new_root == root:
                    raise OSError("could not locate a `pyproject.toml` file")

                root = new_root

        return self._path

    def read(self) -> str:
        with open(self.path, "r", encoding="utf-8") as f:
            return f.read()

    def write(self, text: str):
        with open(self.path, "w", encoding="utf-8") as f:
            f.write(text)

    @abstractmethod
    def load(self) -> Dict[str, Any]:
        """
        Deserializes the pyproject.toml file.
        """

    @abstractmethod
    def save(self, data: Dict[str, Any]):
        """
        Serializes the `data` to the pyproject.toml file.
        """


class StandardHandler(Handler):
    def load(self):
        import tomli

        return tomli.loads(self.read())

    def save(self, data):
        import tomli_w

        self.write(tomli_w.dumps(data))


def get_handler(path: Optional[str] = None):
    # comment-preserving version one day
    return StandardHandler(path)


src/pyproject_validate/models.py
from __future__ import annotations

from typing import Dict, List, Optional, Union

from pydantic import BaseModel, Field, root_validator, validator

# TODO: https://github.com/samuelcolvin/pydantic/issues/1098#issuecomment-1009762620


class LicenseTableLegacy(BaseModel):
    file: Optional[str]
    text: Optional[str]

    @root_validator(pre=True)
    def _pre_validation(cls, values):
        if "file" in values and "text" in values:
            raise ValueError("cannot contain both a `file` and `text` field")

        return values


class AuthorTable(BaseModel):
    email: Optional[str]
    name: Optional[str]


class LicenseFilesTable(BaseModel):
    globs: Optional[List[str]]
    paths: Optional[List[str]]

    @root_validator(pre=True)
    def _pre_validation(cls, values):
        if not ("globs" in values or "paths" in values):
            raise ValueError("must contain either a `globs` or `paths` field")

        return values


class ReadmeTable(BaseModel):
    charset: Optional[str]
    content_type: str = Field(alias="content-type")
    file: Optional[str]
    text: Optional[str]

    @root_validator(pre=True)
    def _pre_validation(cls, values):
        if "file" in values and "text" in values:
            raise ValueError("cannot contain both a `file` and `text` field")

        return values

    @validator("content_type")
    def _validate_content_type(cls, v, field):
        known_content_types = ("text/markdown", "text/x-rst", "text/plain")
        if v not in known_content_types:
            raise ValueError(f'must be one of: {", ".join(known_content_types)}')

        return v


class BuildSystemConfig(BaseModel):
    """
    https://www.python.org/dev/peps/pep-0517/#source-trees
    """

    backend_path: Optional[List[str]] = Field(alias="backend-path")
    build_backend: str = Field(alias="build-backend")
    requires: List[str]


class ProjectConfig(BaseModel):
    """
    https://www.python.org/dev/peps/pep-0621/#details
    """

    authors: Optional[List[AuthorTable]]
    classifiers: Optional[List[str]]
    dependencies: Optional[List[str]]
    description: Optional[str]
    dynamic: Optional[List[str]]
    entry_points: Optional[Dict[str, Dict[str, str]]] = Field(alias="entry-points")
    gui_scripts: Optional[Dict[str, str]] = Field(alias="gui-scripts")
    keywords: Optional[List[str]]
    license: Optional[Union[str, LicenseTableLegacy]]
    license_files: Optional[LicenseFilesTable] = Field(alias="license-files")
    maintainers: Optional[List[AuthorTable]]
    name: str
    optional_dependencies: Optional[Dict[str, List[str]]] = Field(alias="optional-dependencies")
    readme: Optional[Union[str, ReadmeTable]]
    scripts: Optional[Dict[str, str]]
    urls: Optional[Dict[str, str]]
    version: Optional[str]

    @validator("readme")
    def _validate_readme(cls, v, field):
        if not isinstance(v, str):
            return v

        known_extensions = (".md", ".rst", ".txt")
        if not v.lower().endswith(known_extensions):
            raise ValueError(f'must have one of the following extensions: {", ".join(known_extensions)}')

        return v

    @validator("dynamic", pre=False)
    def _validate_dynamic(cls, v, field):
        if v is not None and "name" in v:
            # this would fail later on in the redefined check but let's be explicit
            raise ValueError("the `name` field must not be listed as dynamic")

        return v

    @root_validator(pre=False)
    def _validate_dynamic_fields(cls, values):
        static_fields = dict(values)
        dynamic_fields = set(static_fields.pop("dynamic", None) or [])

        required_fields = ["version"]
        missing = [
            field for field in required_fields if static_fields.get(field) is None and field not in dynamic_fields
        ]
        if missing:
            raise ValueError(f'missing field(s): {", ".join(missing)}')

        redefined = [field for field, value in static_fields.items() if value is not None and field in dynamic_fields]
        if redefined:
            raise ValueError(f'field(s) defined but also listed as dynamic: {", ".join(redefined)}')

        return values
